data/lib/core/achievements.lua

```diff
@@ -0,0 +1,668 @@
+	return targetAchievement
+end
+
+function Player.addAchievement(self, ach, showMsg)
+	local achievement
+	if tonumber(ach) ~= nil then
+		achievement = getAchievementInfoById(ach)
+	else
+		achievement = getAchievementInfoByName(ach)
+	end
+	if not achievement then
+		print("[!] -> Invalid achievement \"" .. ach .. "\".")
+		return false
+	end
+
+	if not self:hasAchievement(achievement.id) then
+		self:setStorageValue(PlayerStorageKeys.achievementsBase + achievement.id, 1)
+		if not showMsg then
+			self:sendTextMessage(MESSAGE_EVENT_ADVANCE, "Congratulations! You earned the achievement \"" .. achievement.name .. "\".")
+		end
+	end
+	return true
+end
+
+function Player.removeAchievement(self, ach)
+	local achievement
+	if tonumber(ach) ~= nil then
+		achievement = getAchievementInfoById(ach)
+	else
+		achievement = getAchievementInfoByName(ach)
+	end
+	if not achievement then
+		print("[!] -> Invalid achievement \"" .. ach .. "\".")
+		return false
+	end
+
+	if self:hasAchievement(achievement.id) then
+		self:setStorageValue(PlayerStorageKeys.achievementsBase + achievement.id, -1)
+	end
+	return true
+end
+
+function Player.addAllAchievements(self, showMsg)
+	for i = ACHIEVEMENT_FIRST, ACHIEVEMENT_LAST do
+		self:addAchievement(i, showMsg)
+	end
+	return true
+end
+
+function Player.removeAllAchievements(self)
+	for k = 1, #achievements do
+		if self:hasAchievement(k) then
+			self:removeAchievement(k)
+		end
+	end
+	return true
+end
+
+function Player.getSecretAchievements(self)
+	local targetAchievement = {}
+	for k, v in pairs(achievements) do
+		if self:hasAchievement(k) and v.secret then
+			targetAchievement[#targetAchievement + 1] = k
+		end
+	end
+	return targetAchievement
+end
+
+function Player.getPublicAchievements(self)
+	local targetAchievement = {}
+	for k, v in pairs(achievements) do
+		if self:hasAchievement(k) and not v.secret then
+			targetAchievement[#targetAchievement + 1] = k
+		end
+	end
+	return targetAchievement
+end
+
+function Player.getAchievementPoints(self)
+	local points = 0
+	local list = self:getAchievements()
+	if #list > 0 then -- has achievements
+		for i = 1, #list do
+			local targetAchievement = getAchievementInfoById(list[i])
+			if targetAchievement.points > 0 then -- avoid achievements with unknow points
+				points = points + targetAchievement.points
+			end
+		end
+	end
+	return points
+end
+
+function Player.addAchievementProgress(self, ach, value)
+	local achievement = tonumber(ach) ~= nil and getAchievementInfoById(ach) or getAchievementInfoByName(ach)
+	if not achievement then
+		print('[!] -> Invalid achievement "' .. ach .. '".')
+		return true
+	end
+
+	local storage = PlayerStorageKeys.achievementsCounter + achievement.id
+	local progress = self:getStorageValue(storage)
+	if progress < value then
+		self:setStorageValue(storage, math.max(1, progress) + 1)
+	elseif progress == value then
+		self:setStorageValue(storage, value + 1)
+		self:addAchievement(achievement.id)
+	end
+	return true
+end
```

Comment:
Nice addition to the code base, I think a few changes can make it even better.

The total amount of times something needs to be completed should be defined in the `achievements` table. A good name for the variable might be `goal`. It doesn't feel right that the `goal` amount needs to be passed into this function for every call. All `goal`s should be defined in the `achievements` table instead of all over the lua scripts.

The last parameter for this function could then me a `count` which is how much progress to add for the achievement towards reaching the goal.

This change makes the system more flexible too. It would be possible to due things like have mana potions add 1 to the progress. strong mana potions add 2 to the progress. ultimate mana potions add 3 to the progress.